### PR TITLE
feat: add cargo_output to eliminate last vestiges of stdout pollution

### PR DESCRIPTION
**Pull request by Gankra**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -30,6 +30,8 @@ libc = { version = "0.2.62", default-features = false, optional = true }
 
 [features]
 parallel = ["dep:libc", "dep:jobserver", "dep:once_cell"]
+# enables OutputKind::Stderr which requires rustc 1.74.0 -- currently above msrv
```
Comment (ChrisDenton) [inline Cargo.toml:33]: Hm, maybe it's time to bump the MSRV to 1.75 (RHEL 8.10 & 9.4, Ubuntu 24.04LTS seem to have settled on that version).

**Comment by Gankra** [comment]
context: abi-cafe uses cc-rs as a library, and *really* wants stdout to be totally clean for its own output report (which may be json, so, very sad if that's messed with).

2 years ago I forked cc-rs to remove all printlns and redirect subcommand stderr to piped (effectively null): https://github.com/Gankra/abi-cafe/issues/49

Since then y'all have wonderfully added the `cargo_*` APIs to address this exact problem, *however* the "stderr to null" case is still outstanding. On the happy path compilers don't write to stdout, but when the source code has errors they like spewing them to stdout!

Of course I don't *really* want to lose those errors, so I included a "pipe stdout to stderr" mode -- unfortunately violating current cc msrv.

**Comment by Gankra** [comment]
Usage of this PR in abi-cafe: https://github.com/Gankra/abi-cafe/pull/58

**Comment by Gankra** [comment]
I will also concede that this PR is still technically a half-measure. What I *reaaally* want is a way to capture all compiler output (stderr *and* stdout) to a String(s?) (or `Vec<u8>`, whatevs) when invoking `try_compile`. I briefly experimented with what this would look like and it was a bit too messy for my inexperience with this codebase.

File: Cargo.toml (hunk 1 above)
Comment (Gankra) [reply Cargo.toml:33]: This would certainly rock for me personally

**Comment by Gankra** [comment]
I also want to make it clear the Stderr mode isn't a blocker for me -- I was already blackholing stdout in my fork, I just know this std trick was added since then and think it's nice to have.

File: Cargo.toml (hunk 1 above)
Comment (thomcc) [reply Cargo.toml:33]: [Elsewhere](https://github.com/rust-lang/cc-rs/pull/1140#issuecomment-2213008655) I've recommended we switch to stable-5 (which I believe is in line with `libc`'s stated MSRV... even if it's actual MSRV is preposterously old) which would give us 1.74.

Comment (thomcc) [reply Cargo.toml:33]: Just to verify, the only reason this is behind a feature is the MSRV situation, right?

**Review by thomcc — COMMENTED** [review]
This looks fine to me. I think we should consider bumping MSRV to 1.74 for this. Removing a cargo feature is considered a breaking change I believe, so I'd prefer to bump the MSRV then to add a cargo feature.

This would be an aggressive enough bump that it might be worth giving a heads up in t-libs/crate-maintainers zulip first tho.

**Comment by thomcc** [comment]
> I also want to make it clear the Stderr mode isn't a blocker for me -- I was already blackholing stdout in my fork, I just know this std trick was added since then and think it's nice to have.

Oh, I might prefer to wait on it then.

File: Cargo.toml (hunk 1 above)
Comment (Gankra) [reply Cargo.toml:33]: correct

**Comment by Gankra** [comment]
Also if we're cutting down to that I suppose we can make `cargo_output` take a bool and internally desugar to these options?

**Comment by Gankra** [comment]
(It's forward compat to add enum to the public API later, with an `impl Into<OutputKind> for bool` kinda thing

**Comment by NobodyXu** [comment]
> Ok so would we be happy landing this with the stderr stuff removed?

Yeah, alternatively, perhaps we can take a `Box<dyn (Into<Stdio> + Clone + Send + Sync) + 'static>` instead?

**Comment by Gankra** [comment]
Yeah i played around with a sufficiently fancy `dyn` but you can't have + Clone or + Debug in there (error[E0225]: only auto traits can be used as additional traits in a trait object)

**Comment by NobodyXu** [comment]
> + Clone

Aha wait I see the problem with `Clone`, the `Self` type is the problem...

So maybe we could take a function `Box<dyn Fn() -> Stdio + Send + Sync + 'static>`?

It's a bit hacky though...

**Review by thomcc — APPROVED** [review]
Looks good to me.

**Comment by thomcc** [comment]
I think we should take this for now, and once we have a new enough MSRV, (try to remember to) extend the API later with the more advanced functionality rather than hacking it in.

**Comment by workingjubilee** [comment]
> This looks fine to me. I think we should consider bumping MSRV to 1.74 for this. Removing a cargo feature is considered a breaking change I believe, so I'd prefer to bump the MSRV then to add a cargo feature.

To clarify, removing a cargo feature and releasing a version without that feature means that dependees using that feature do *not* upgrade to newer versions.

**Comment by Gankra** [comment]
(also to clarify i would have suggested the feature grows to be defaulted on once msrv is achieved but this is all moot now)

**Pull request merged**